### PR TITLE
added support for bedrock/claude3

### PR DESCRIPTION
**Pull request by koid**
## **User description**
added support for bedrock/claude3

___

## **Type**
enhancement, documentation

___

## **Description**
- Added support for new Bedrock Claude3 models in the AI handlers.
- Introduced a new configuration option `drop_params` to ensure compatibility with Claude3 models.
- Updated the documentation to guide users on configuring Claude3 models and the new `drop_params` setting.
- Updated the `litellm` library version to `1.31.10` for enhanced functionality and compatibility.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>__init__.py</strong><dd><code>Add Support for Bedrock Claude3 Models</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pr_agent/algo/__init__.py

<li>Added support for <code>bedrock/anthropic.claude-3-sonnet-20240229-v1:0</code> and <br><code>bedrock/anthropic.claude-3-haiku-20240307-v1:0</code> models.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/781/files#diff-5923c546f24ec7308a0e43fc84bb6fe40de7bfe2ac6ee842da9578e5dc2c692b">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>litellm_ai_handler.py</strong><dd><code>Enhance Litellm AI Handler for Claude3 Compatibility</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pr_agent/algo/ai_handlers/litellm_ai_handler.py

<li>Added condition to drop incompatible parameters when using Claude3 <br>models.<br> <li> Updated <code>max_tokens</code> for <code>AmazonAnthropicClaude3Config</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/781/files#diff-ea1acaa0907f3410665530fbc4cda2ab524de2772e0bbe10bad4648b8be35dfe">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>configuration.toml</strong><dd><code>Add Drop Params Configuration Option</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pr_agent/settings/configuration.toml

- Added `drop_params` configuration option in `litellm` section.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/781/files#diff-66cfda5143e484ee53ecf7aa0df7dca8ad0b181256f4b0675905db35bcbbae78">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Documentation
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>additional_configurations.md</strong><dd><code>Update Documentation for Claude3 Model Configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
docs/docs/usage-guide/additional_configurations.md

<li>Updated documentation to include configuration for Claude3 models.<br> <li> Added note on setting <code>drop_params</code> for Claude3 compatibility.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/781/files#diff-9290a3ad9a86690b31f0450b77acd37ef1914b41fabc8a08682d4da433a77f90">+9/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Dependencies
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>requirements.txt</strong><dd><code>Update Litellm Library Version</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
requirements.txt

- Updated `litellm` library version from `1.29.1` to `1.31.10`.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/781/files#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: docs/docs/usage-guide/additional_configurations.md [hunk 1]
```diff
@@ -162,15 +162,22 @@ To use Amazon Bedrock and its foundational models, add the below configuration:
...
+model_turbo="bedrock/anthropic.claude-3-sonnet-20240229-v1:0"
+fallback_models=["bedrock/anthropic.claude-v2:1"]
 
 [aws] # in .secrets.toml
 bedrock_region = "us-east-1"
 ```
 
 Note that you have to add access to foundational models before using them. Please refer to [this document](https://docs.aws.amazon.com/bedrock/latest/userguide/setting-up.html) for more details.
 
+If you are using the claude-3 model, please configure the following settings as there are parameters incompatible with claude-3.
```
Comment (koid) [inline docs/docs/usage-guide/additional_configurations.md:175]: bedrock does not support `temperature` parameter

https://github.com/BerriAI/litellm/blob/1b93ada827c652eed5f4d8ac05b222b10222b80d/litellm/llms/bedrock.py#L79-L87

File: pr_agent/algo/ai_handlers/litellm_ai_handler.py [hunk 2]
```diff
@@ -68,6 +70,7 @@ def __init__(self):
             )
         if get_settings().get("AWS.BEDROCK_REGION", None):
             litellm.AmazonAnthropicConfig.max_tokens_to_sample = 2000
+            litellm.AmazonAnthropicClaude3Config.max_tokens = 2000
```
Comment (koid) [inline pr_agent/algo/ai_handlers/litellm_ai_handler.py:73]: In litellm, a separate Config class is created for AmazonAnthropicClaude3

Similar to the following pull request, the value of max_tokens is hardcoded.
https://github.com/Codium-ai/pr-agent/pull/483#discussion_r1407657224

File: docs/docs/usage-guide/additional_configurations.md (hunk 1 above)
Comment (koid) [reply docs/docs/usage-guide/additional_configurations.md:175]: Discussion:
Another approach could be to set the temperature only if the model is bedrock's claude3 within the chat_completion function.

File: pr_agent/algo/ai_handlers/litellm_ai_handler.py (hunk 2 above)
Comment (koid) [reply pr_agent/algo/ai_handlers/litellm_ai_handler.py:73]: Discussion:
Another approach within chat_completion is to handle which configuration to set based on the model, either AmazonAnthropicConfig or AmazonAnthropicClaude3Config.

**Comment by mrT23** [comment]
Hi @koid , thanks for the PR

Is the upgrade for litellm version essential for this PR ?
It sometimes creates problems and conflicts, and I would like to avoid it if possible.
Otherwise, looks good.

As a side note, did you form an impression of Claude-3 opus, compared to GPT4, on code? is it competitive ?

**Comment by koid** [comment]
Hi @mrT23 , thanks for review

The support status for litellm's Claude-3 is as follows, and an update is required.
- haiku from version [v1.31.9](https://github.com/BerriAI/litellm/releases/tag/v1.31.9)
- sonnet from version [v1.29.3](https://github.com/BerriAI/litellm/releases/tag/v1.29.3)

---

I'm sorry, I haven't had the chance to try Claude-3 Opus yet. I plan to try it soon.
I haven't done a strict comparison, but this is my personal opinion:
Sonnet is at a usable level and has advantages in terms of cost and response speed.
I feel that Haiku is at a similar level to GPT-3.5.

**Comment by mrT23** [comment]
Preparing review...

**Comment by mrT23** [comment]
## PR Review

<table>
<tr>
<tr><td> ⏱️&nbsp;<strong>Estimated&nbsp;effort&nbsp;to&nbsp;review [1-5]</strong></td><td>

2, because the changes are straightforward and mostly involve configuration updates and minor code adjustments. The documentation and requirements update are clear and well-defined.


</td></tr>
<tr><td> 🧪&nbsp;<strong>Relevant tests</strong></td><td>

No


</td></tr>
<tr><td> 🔍&nbsp;<strong>Possible issues</strong></td><td>

<strong>Possible Bug:</strong> The addition of `litellm.drop_params` in `litellm_ai_handler.py` without checking its type or existence could lead to runtime errors if not properly configured.

</td></tr>
<tr><td> 🔒&nbsp;<strong>Security concerns</strong></td><td>

No

</td></tr>
</table>


<details><summary> <strong>Code feedback:</strong></summary>

<hr><table><tr><td>relevant file</td><td>pr_agent/algo/ai_handlers/litellm_ai_handler.py
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Consider adding a type check for `litellm.drop_params` to ensure it's a boolean. This prevents potential runtime errors if the configuration is incorrect. [important]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/Codium-ai/pr-agent/pull/781/files#diff-ea1acaa0907f3410665530fbc4cda2ab524de2772e0bbe10bad4648b8be35dfeR40'>litellm.drop_params = get_settings().litellm.drop_params</a></td></tr></table><hr>

<table><tr><td>relevant file</td><td>pr_agent/algo/__init__.py
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Ensure that the model keys added are consistent and correctly formatted. Double-check for any typos or inconsistencies in model naming conventions. [medium]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/Codium-ai/pr-agent/pull/781/files#diff-5923c546f24ec7308a0e43fc84bb6fe40de7bfe2ac6ee842da9578e5dc2c692bR29'>'bedrock/anthropic.claude-3-sonnet-20240229-v1:0': 100000,</a></td></tr></table><hr>

<table><tr><td>relevant file</td><td>docs/docs/usage-guide/additional_configurations.md
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Add a note about the expected type for `drop_params` in the documentation to guide users on proper configuration. [medium]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/Codium-ai/pr-agent/pull/781/files#diff-9290a3ad9a86690b31f0450b77acd37ef1914b41fabc8a08682d4da433a77f90R178'>drop_params = true</a></td></tr></table><hr>

<table><tr><td>relevant file</td><td>requirements.txt
</td></tr><tr><td>suggestion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</td><td>

<strong>

Verify the compatibility of the updated `litellm` library version with the existing codebase to avoid potential integration issues. [important]

</strong>
</td></tr><tr><td>relevant line</td><td><a href='https://github.com/Codium-ai/pr-agent/pull/781/files#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552R12'>litellm==1.31.10</a></td></tr></table><hr>

</details><hr>

<details> <summary><strong>✨ Review tool usage guide:</strong></summary><hr> 

**Overview:**
The `review` tool scans the PR code changes, and generates a PR review. The tool can be triggered [automatically](https://pr-agent-docs.codium.ai/usage-guide/automations_and_usage/#github-app-automatic-tools-when-a-new-pr-is-opened) every time a new PR is opened, or can be invoked manually by commenting on any PR.
When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L19) related to the review tool (`pr_reviewer` section), use the following template:
```
/review --pr_reviewer.some_config1=... --pr_reviewer.some_config2=...
```
With a [configuration file](https://pr-agent-docs.codium.ai/usage-guide/configuration_options/), use the following template:
```
[pr_reviewer]
some_config1=...
some_config2=...
```
    

<table><tr><td><details> <summary><strong> Utilizing extra instructions</strong></summary><hr>

The `review` tool can be configured with extra instructions, which can be used to guide the model to a feedback tailored to the needs of your project.

Be specific, clear, and concise in the instructions. With extra instructions, you are the prompter. Specify the relevant sub-tool, and the relevant aspects of the PR that you want to emphasize.

Examples for extra instructions:
```
[pr_reviewer] # /review #
extra_instructions="""
In the 'possible issues' section, emphasize the following:
- Does the code logic cover relevant edge cases?
- Is the code logic clear and easy to understand?
- Is the code logic efficient?
...
"""
```
Use triple quotes to write multi-line instructions. Use bullet points to make the instructions more readable.
        

</details></td></tr>

<tr><td><details> <summary><strong> How to enable\disable automation</strong></summary><hr>

- When you first install PR-Agent app, the [default mode](https://pr-agent-docs.codium.ai/usage-guide/automations_and_usage/#github-app-automatic-tools-when-a-new-pr-is-opened) for the `review` tool is:
```
pr_commands = ["/review", ...]
```
meaning the `review` tool will run automatically on every PR, with the default configuration.
Edit this field to enable/disable the tool, or to change the used configurations
        

</details></td></tr>

<tr><td><details> <summary><strong> Auto-labels</strong></summary><hr>

The `review` tool can auto-generate two specific types of labels for a PR:
- a `possible security issue` label, that detects possible [security issues](https://github.com/Codium-ai/pr-agent/blob/tr/user_description/pr_agent/settings/pr_reviewer_prompts.toml#L136) (`enable_review_labels_security` flag)
- a `Review effort [1-5]: x` label, where x is the estimated effort to review the PR (`enable_review_labels_effort` flag)


</details></td></tr>

<tr><td><details> <summary><strong> Extra sub-tools</strong></summary><hr>

The `review` tool provides a collection of possible feedbacks about a PR.
It is recommended to review the [possible options](https://pr-agent-docs.codium.ai/tools/review/#enabledisable-features), and choose the ones relevant for your use case.
Some of the feature that are disabled by default are quite useful, and should be considered for enabling. For example: 
`require_score_review`, `require_soc2_ticket`, `require_can_be_split_review`, and more.


</details></td></tr>

<tr><td><details> <summary><strong> Auto-approve PRs</strong></summary><hr>

By invoking:
```
/review auto_approve
```
The tool will automatically approve the PR, and add a comment with the approval.


To ensure safety, the auto-approval feature is disabled by default. To enable auto-approval, you need to actively set in a pre-defined configuration file the following:
```
[pr_reviewer]
enable_auto_approval = true
```
(this specific flag cannot be set with a command line argument, only in the configuration file, committed to the repository)


You can also enable auto-approval only if the PR meets certain requirements, such as that the `estimated_review_effort` is equal or below a certain threshold, by adjusting the flag:
```
[pr_reviewer]
maximal_review_effort = 5
```


</details></td></tr>



<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://pr-agent-docs.codium.ai/tools/) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 


</details></td></tr>

</table>

See the [review usage](https://pr-agent-docs.codium.ai/tools/review/) page for a comprehensive guide on using this tool.


</details>

**Review by mrT23 — APPROVED** [review]

**Pull request merged**